### ISSUE TITLE
Multi-value params parsing fails when no end marker found

### ISSUE THREAD

**Issue by ikorchynskyi** [shlomif/perl-Config-IniFiles#6]
STR:
```
$ perl -Mstrict -MConfig::IniFiles -lwe 'my $data = qq{[A]\nb=<<c\nd}; Config::IniFiles->new( -file => \$data ) or print @Config::IniFiles::errors;'
Can't use an undefined value as a symbol reference at /home/user/perl5/lib/perl5/Config/IniFiles.pm line 495.
```

It seems, that `$fh` should be passed in `_rollback` at line 745, to get expected `2: no end marker ("c") found` error message without dying:
```
 716 sub _ReadConfig_handle_here_doc_param
 717 {
 718     my ( $self, $fh, $eotmark, $val_aref ) = @_;
 719 
 720     my $foundeot  = 0;
 721     my $startline = $self->_read_line_num();
 722 
 723 HERE_DOC_LOOP:
 724     while ( defined( my $line = $self->_read_next_line($fh) ) )
 725     {
 726         if ( $line eq $eotmark )
 727         {
 728             $foundeot = 1;
 729             last HERE_DOC_LOOP;
 730         }
 731         else
 732         {
 733             # Untaint
 734             my ($contents) = $line =~ /(.*)/ms;
 735             CORE::push( @$val_aref, $contents );
 736         }
 737     }
 738 
 739     if ( !$foundeot )
 740     {
 741         $self->_add_error(
 742             sprintf( '%d: %s',
 743                 $startline, qq#no end marker ("$eotmark") found# )
 744         );
 745         $self->_rollback();
 746         return $RET_BREAK;
 747     }
 748 
 749     return $RET_CONTINUE;
 750 }
```

**Comment by shlomif** [comment]
On Tue, 12 Mar 2019 16:06:32 +0000 (UTC)
Igor Korchynskyi <notifications@github.com> wrote:

> STR:
> ```
> $ perl -Mstrict -MConfig::IniFiles -lwe 'my $data = qq{[A]\nb=<<c\nd};
> Config::IniFiles->new( -file => \$data ) or print @Config::IniFiles::errors;'
> Can't use an undefined value as a symbol reference at
> /home/user/perl5/lib/perl5/Config/IniFiles.pm line 495. ```
> 

Thanks, Igor! Can you prepare a pull-req for fixing this? (with a t/*.t file)

> It seems, that `$fh` should be passed in `_rollback` at line 745, to get
> expected `2: no end marker ("c") found` error message without dying: ```
>  716 sub _ReadConfig_handle_here_doc_param
>  717 {
>  718     my ( $self, $fh, $eotmark, $val_aref ) = @_;
>  719 
>  720     my $foundeot  = 0;
>  721     my $startline = $self->_read_line_num();
>  722 
>  723 HERE_DOC_LOOP:
>  724     while ( defined( my $line = $self->_read_next_line($fh) ) )
>  725     {
>  726         if ( $line eq $eotmark )
>  727         {
>  728             $foundeot = 1;
>  729             last HERE_DOC_LOOP;
>  730         }
>  731         else
>  732         {
>  733             # Untaint
>  734             my ($contents) = $line =~ /(.*)/ms;
>  735             CORE::push( @$val_aref, $contents );
>  736         }
>  737     }
>  738 
>  739     if ( !$foundeot )
>  740     {
>  741         $self->_add_error(
>  742             sprintf( '%d: %s',
>  743                 $startline, qq#no end marker ("$eotmark") found# )
>  744         );
>  745         $self->_rollback();
>  746         return $RET_BREAK;
>  747     }
>  748 
>  749     return $RET_CONTINUE;
>  750 }
> ```
> 



-- 
-----------------------------------------------------------------
Shlomi Fish       http://www.shlomifish.org/
Apple Inc. is Evil - http://www.shlomifish.org/open-source/anti/apple/

The reward for a job well done will generally just be another, probably harder,
job.
    — Anthony Towns

Please reply to list if it's a mailing list post - http://shlom.in/reply .

**Comment by ikorchynskyi** [comment]
> Can you prepare a pull-req for fixing this? (with a t/*.t file)

Prepared #7

**Comment by ikorchynskyi** [comment]
Seems to be merged & fixed, 3.000002 released on CPAN.
